File: old/algo/316.kt
// class Solution {
//     fun removeDuplicateLetters(s: String): String {
//         val sSet = s.toSet()
//         for (ch in sSet.sorted()) {
//             val suffix = s.substring(s.indexOf(ch))
//             if (sSet == suffix.toSet()) {
//                 return ch + removeDuplicateLetters(suffix.replace(ch.toString(), ""))
//             }
//         }
//         return ""
//     }
// }

// //val s = "bcabc"
// //val s = "bcdabc"
// val s = "cbacdcbc"

println(Solution().removeDuplicateLetters(s))

class Solution {
    fun removeDuplicateLetters(s: String): String {
        val seen = mutableSetOf<Char>()
        val counts = HashMap<Char, Int>()
        for (ch in s) {
            counts[ch] = counts.getOrDefault(ch, 0) + 1
        }

        val stack = mutableListOf<Char>()
        for (ch in s) {
            counts[ch] = counts.getOrDefault(ch, 0) - 1 // 디폴트 값이 나올 수 없음.
            if (seen.contains(ch)) continue

            while (stack.isNotEmpty() && stack[stack.lastIndex] > ch && counts.getOrDefault(stack[stack.lastIndex], 0) > 0) {
                val lastCh = stack.removeAt(stack.lastIndex)
                seen.remove(lastCh)
            }

            stack.add(ch)
            seen.add(ch)
        }
        return stack.joinToString("")
    }
}

//val s = "bcabc"
//val s = "bcdabc"
val s = "cbacdcbc"

println(Solution().removeDuplicateLetters(s))

File: old/algo/316.py
import collections

# 46ms 재귀 방식

# class Solution:
#     def removeDuplicateLetters(self, s: str) -> str:
#         set_s = set(s)

#         for i in sorted(set_s):
#             suffix = s[s.index(i):]
#             if set_s == set(suffix):
#                 return i + self.removeDuplicateLetters(suffix.replace(i, ''))
#         return ''

# 34ms set와 Counter를 이용한 방식
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        counter, stack, seen = collections.Counter(s), [], set()
        for char in s:
            counter[char] -= 1
            if char in seen:
                continue
            while stack and char < stack[-1] and counter[stack[-1]]:
                seen.remove(stack.pop())                
            stack.append(char)
            seen.add(char)
        return ''.join(stack)

a = Solution()

s = "bcabc"
print(a.removeDuplicateLetters(s))

s = "cbacdcbc"
print(a.removeDuplicateLetters(s))
